pkg/keycloak/realm/phaseHandler.go

```diff
@@ -0,0 +1,449 @@
+package realm
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/aerogear/keycloak-operator/pkg/apis/aerogear/v1alpha1"
+	"github.com/aerogear/keycloak-operator/pkg/keycloak"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type phaseHandler struct {
+	k8sClient       kubernetes.Interface
+	sdk             keycloak.SdkCruder
+	operatorNS      string
+	kcClientFactory keycloak.KeycloakClientFactory
+	defaultClients  map[string]struct{}
+}
+
+func NewPhaseHandler(k8sClient kubernetes.Interface, sdk keycloak.SdkCruder, operatorNS string, kcFactory keycloak.KeycloakClientFactory) *phaseHandler {
+	kcDefaultClients := []string{"account", "admin-cli", "broker", "realm-management", "security-admin-console"}
+	set := make(map[string]struct{}, len(kcDefaultClients))
+	for _, s := range kcDefaultClients {
+		set[s] = struct{}{}
+	}
+	return &phaseHandler{
+		k8sClient:       k8sClient,
+		sdk:             sdk,
+		operatorNS:      operatorNS,
+		kcClientFactory: kcFactory,
+		defaultClients:  set,
+	}
+}
+
+func (ph *phaseHandler) Initialise(kcr *v1alpha1.KeycloakRealm) (*v1alpha1.KeycloakRealm, error) {
+	// set the finalizer
+	if err := v1alpha1.AddFinalizer(kcr, v1alpha1.KeycloakFinalizer); err != nil {
+		return nil, err
+	}
+	// set the phase to accepted or set a message that it cannot be accepted
+	kcr.Status.Phase = v1alpha1.PhaseAccepted
+	return kcr, nil
+}
+
+func (ph *phaseHandler) Accepted(kcr *v1alpha1.KeycloakRealm) (*v1alpha1.KeycloakRealm, error) {
+	//look for a provisioned keycloak instance
+	list := &v1alpha1.KeycloakList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "keycloak",
+			APIVersion: v1alpha1.Group + "/" + v1alpha1.Version,
+		},
+	}
+	ph.sdk.List(ph.operatorNS, list, nil)
+
+	for _, kc := range list.Items {
+		if kc.Status.Phase == v1alpha1.PhaseAwaitProvision {
+			kc.Status.Phase = v1alpha1.PhaseProvision
+			ph.sdk.Update(&kc)
+		}
+		if kc.Status.Phase == v1alpha1.PhaseComplete {
+			kcr.Status.KeycloakName = kc.Name
+			kcr.Status.Phase = v1alpha1.PhaseProvision
+		}
+	}
+	return kcr, nil
+}
+
+func (ph *phaseHandler) Provision(kcr *v1alpha1.KeycloakRealm) (*v1alpha1.KeycloakRealm, error) {
+	kcClient, err := ph.getClient(kcr)
+	if err != nil {
+		return kcr, err
+	}
+
+	realm, err := kcClient.GetRealm(kcr.Name)
+	if err != nil {
+		return kcr, errors.Wrap(err, "error retrieving realms from keycloak")
+	}
+
+	if realm != nil && kcr != nil && realm.Spec.Realm == kcr.Spec.Realm {
+		kcr.Status.Phase = v1alpha1.PhaseProvisioned
+		return kcr, nil
+	}
+
+	storeUsers := kcr.Spec.KeycloakApiRealm.Users
+	storeClients := kcr.Spec.KeycloakApiRealm.Clients
+
+	kcr.Spec.KeycloakApiRealm.Clients = []v1alpha1.KeycloakClient{}
+	kcr.Spec.KeycloakApiRealm.Users = []v1alpha1.KeycloakUser{}
+
+	err = kcClient.CreateRealm(kcr)
+
+	kcr.Spec.KeycloakApiRealm.Clients = storeClients
```

Comment:
because they are not keycloakApiUser object, they are keycloakUser objects, so when they get converted to JSON and sent to the keycloak API they have values (like outputSecret) that cause the API create endpoint to fail.

I decided the simplest approach was to remove the users from the create request, and allow the reconciliation of users and clients to correct this.